specs/lazy-adr/adr-010-sequencing.md

```diff
@@ -0,0 +1,240 @@
+	// returns the next batch of transactions and an error if any from the sequencer
+	GetNextBatch(ctx context.Context, req GetNextBatchRequest) (*GetNextBatchResponse, error)
+}
+
+// BatchVerifier provides a method for verifying a batch of transactions received from the sequencer
+type BatchVerifier interface {
+	// VerifyBatch verifies a batch of transactions received from the sequencer
+	// RollupId is the unique identifier for the rollup chain
+	// BatchHash is the cryptographic hash of the batch to verify
+	// returns a boolean indicating if the batch is valid and an error if any from the sequencer
+	VerifyBatch(ctx context.Context, req VerifyBatchRequest) (*VerifyBatchResponse, error)
+}
+
+// Batch is a collection of transactions
+type Batch struct {
+	Transactions [][]byte
+}
+
+// SubmitRollupTransactionRequest is a request to submit a transaction from rollup to sequencer
+type SubmitRollupTransactionRequest struct {
+	RollupId []byte
+	Tx       []byte
+}
+
+// SubmitRollupTransactionResponse is a response to submitting a transaction from rollup to sequencer
+type SubmitRollupTransactionResponse struct {
+}
+
+// GetNextBatchRequest is a request to get the next batch of transactions from sequencer to rollup
+type GetNextBatchRequest struct {
+	RollupId      []byte
+	LastBatchHash []byte
+	MaxBytes      uint64
+}
+
+// GetNextBatchResponse is a response to getting the next batch of transactions from sequencer to rollup
+type GetNextBatchResponse struct {
+	Batch     *Batch
+	Timestamp time.Time
+}
+
+// VerifyBatchRequest is a request to verify a batch of transactions received from the sequencer
+type VerifyBatchRequest struct {
+	RollupId  []byte
+	BatchHash []byte
+}
+
+// VerifyBatchResponse is a response to verifying a batch of transactions received from the sequencer
+type VerifyBatchResponse struct {
+	Status bool
+}
+```
+
+## Alternative Approaches
+
+The current sequencing interface lacks the following features:
+* The design and interface is not clear how the rollkit node types (header producers and full nodes) are supposed to use it
+* The verification (VerifyBatch) is completely omitted
+* The last batch hash is not sufficient to refer to the previous batch while producing or fetching the next batch
+
+## Decision
+
+The proposed sequencing interface is as follows:
+
+```
+// Sequencer is a generic interface for a rollup sequencer
+type Sequencer interface {
+	SequencerInput
+	SequencerOutput
+	BatchVerifier
+}
+
+// SequencerInput provides a method for submitting a transaction from rollup to sequencer
+type SequencerInput interface {
+	// SubmitRollupTransactions submits a list of transactions from rollup to sequencer
+	// RollupId is the unique identifier for the rollup chain
+	// Tx is the transaction to submit
+	// returns an error if any from the sequencer
+	SubmitRollupTransactions(ctx context.Context, req SubmitRollupTransactionRequest) (*SubmitRollupTransactionResponse, error)
```

Comment:
we should do batches here, there was an open issue in go-sequencing so i made the change in the exec-api branch

Reply:
yes `SubmitRollupTransactionRequest` is modified to do `Txs` instead of `Tx`.
```
type SubmitRollupTransactionRequest struct {
	RollupId []byte
	Txs      [][]byte
}
```